Blog posts for 1.0 release (#15)
Open-EO/openeo.org: news/2020-07-31-openeo-api-released.md

---
news: true
title:  Stable release of the openEO API and Processes (v1.0.0)
date:   2020-07-31
author: Matthias Mohr
---

> After two release candidates in early 2020, the openEO Consortium released the new first stable version of its [openEO API](https://api.openeo.org/) and their set of [common processes](https://processes.openeo.org). The following blog post will give a short overview over the most notable changes since [version 0.4](2019-03-07-openeo-api-040.md) and provide you with an outlook into the next months.

Reviewer: ```suggestion
After two release candidates in early 2020, the openEO Consortium released the first stable version of its [openEO API](https://api.openeo.org/) and their set of [common processes](https://processes.openeo.org). This blog post gives an overview of the most notable changes since [version 0.4](2019-03-07-openeo-api-040.md) and provides an outlook on the next months' project activity.
```

  
## Processes

> The most important change was introduced just recently in openEO API version 1.0.0 RC2. Until then, openEO process graphs allowed users to express processing algorithms and execute them on a back-end using *pre-defined processes* exposed by a back-end provider. While this still exists, a user can also promote a process graph to a full openEO process description, following the same schema as the pre-defined processes. So it includes the processing instructions and all the metadata to make it useful, like an identifier, parameters, return values, examples etc. This allows *user-defined processes* to be stored and used like pre-defined processes, which makes extending the processing capabilities much easier for users. They can now encapsulate algorithms into separate user-defined processes and re-use them in other user-defined processes. It also allows to exchange processes easily between users and back-ends. This offers the possibility of a repository of user-contributed processes that other users can also benefit from. The place to find and share such processes is planned to be the [openEO Hub](https://hub.openeo.org).

Reviewer: ```suggestion
The most important change was introduced just recently in openEO API version 1.0.0 RC2. Until then, openEO process graphs allowed users to express processing algorithms and execute them on a back-end using *pre-defined processes* exposed by a back-end provider. While this still exists, a user can now also promote a process graph to a full openEO process description, following the same schema as the pre-defined processes: it must include the processing instructions and metadata like an identifier, parameters, return values, examples etc. This allows *user-defined processes* to be stored and used like pre-defined processes, enabling users to extend the processing capabilities of a back-end. Users can now encapsulate algorithms in separate user-defined processes and re-use these in other user-defined processes. This also allows the exchange of processes (algorithms) between users and back-ends, and creates the possibility for a repository of user-contributed processes that other users can benefit from. A place for finding and sharing such processes is (planned to be) the [openEO Hub](https://hub.openeo.org).
```

> Through user-defined processes (and user defined functions) users may also be able to fill gaps of processes that back-ends have not implemented (yet). In fact several pre-defined processes are defined with alternative processing instructions that use other openEO pre-defined processes to solve the task. For example, the pre-defined process `normalized_ difference` can be emulated with a combination of the processes `add`, `subtract` and `divide`. So if a back-end is missing the `normalized_difference` process, just [download it](https://processes.openeo.org/#normalized_difference) and use our clients to push it to the back-end and then they act as drop-in replacement for the missing pre-defined process. In the future we plan to allow just using the URL of the process, so no steps in-between would be required.

Reviewer: ```suggestion
Through user-defined processes (and user defined functions) users may also be able to substitute processes that back-ends have not implemented. In fact, several pre-defined processes are defined with alternative processing instructions that use other openEO pre-defined processes to solve the task. For example, the pre-defined process `normalized_ difference` can be emulated with a combination of the processes `add`, `subtract` and `divide`. So if a back-end is missing the `normalized_difference` process, a user can just [download it](https://processes.openeo.org/#normalized_difference) and use an openEO client to push it to the back-end where it then acts as drop-in replacement for the missing pre-defined process. In the future we plan to allow just using the URL of the process, so that no download/upload would be required.
```

> As mentioned above, openEO offers a set of [common processes](https://processes.openeo.org) to improve interoperability and portability between back-ends. This is not a complete list and is expected to be extended by both the openEO Consortium. More than that, we'd like to see that processes arise from the community and that we can take over and standardize whatever comes up and is useful for the different domains working with openEO. We can then simply use the user-defined processes as they share the same schema as the pre-defined processes.

Reviewer: ```suggestion
As mentioned above, openEO offers a set of [common processes](https://processes.openeo.org) to improve interoperability and portability between back-ends. This is not a complete list and is expected to be extended by the openEO Consortium. More than that, we would like to see that processes arise from the community and that we can take over and standardize whatever comes up and is useful for the different domains working with openEO. We can then simply use the user-defined processes as they share the same schema as the pre-defined processes.
```

## Other improvements

Other notable changes and additions:

* Support for importing user-provided data sets has been added.
* Providers can now expose their terms of service and privacy policies.
* The WebSocket-based Subscription API for notifications and monitoring introduced in version 0.4 has been replaced by a logging mechanism in the HTTP API. This makes implementing the API easier for clients and back-ends, but doesn't work in real-time any longer.

The full set of changes can be found in the [API's change log](https://github.com/Open-EO/openeo-api/blob/1.0.0/CHANGELOG.md).
  
## Limitations

With openEO, we are digging into new grounds regarding the interoperability of big Earth observation clouds. While we could solve a lot of issues that came up, there were some issues that we couldn't solve as part of the API and processes. We want to be honest and give some insights into the lessons learned, so I have posted these as a [separate blog post](2020-07-31-lessons-learned.md).

## Next steps

> The API's development went through several iterations and is working well for our broad range of use cases, which we'll report on soon. [Implementations](/software.md) by several organizations for several API versions provide evidence that the API and the processes are useful for a wide range of use cases. Client and back-end developers will now finish their implementations and we hope to have the first operational services this year. We will follow up with new blog posts once other releases are available. If you are interested in trying out openEO, please [contact us](/contact.md) to get more information.

Reviewer: ```suggestion
The API's development went through several iterations and is working well for our broad range of use cases, which we'll report soon. [Implementations](/software.md) by several organizations for several API versions provide evidence that the API and the processes are useful for a wide range of use cases. Client and back-end developers will now finish their implementations and we hope to have the first operational services this year. We will follow up with new blog posts once other releases are available. If you are interested in trying out openEO, please [contact us](/contact.md) to get more information.
```
